src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java

```diff
@@ -0,0 +1,51 @@
+package org.commcare.formplayer.services;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.commcare.formplayer.screens.FormplayerQueryScreen;
+import org.commcare.modern.util.Pair;
+import org.javarosa.core.model.instance.ExternalDataInstance;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cache.CacheManager;
+import org.springframework.cache.annotation.CacheConfig;
+import org.springframework.cache.annotation.CacheEvict;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.lang.Nullable;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.web.client.RestTemplate;
+
+import java.io.ByteArrayInputStream;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+
+@CacheConfig(cacheNames = {"case_search"})
+public class CaseSearchHelper {
+
+    @Autowired
+    CacheManager cacheManager;
+
+    @Autowired
+    QueryRequester queryRequester;
+
+    @Cacheable(unless="#result == null", key="#username + #uri")
+    public ExternalDataInstance getSearchDataInstance(String username, FormplayerQueryScreen screen, String uri, HttpHeaders headers) {
+        String responseString = queryRequester.makeQueryRequest(uri, headers);
+        if (responseString != null) {
+            Pair<ExternalDataInstance, String> dataInstanceWithError = screen.processResponse(
+                    new ByteArrayInputStream(responseString.getBytes(StandardCharsets.UTF_8)));
+            return dataInstanceWithError.first;
```

Comment:
One thing about the `ExternalDataInstance`: I *don't* think it's safe, unfortunately, for us to reuse the raw java object, since it's not built for concurrency. 

@snopoke Is there a way to manually tell the system HOW to cache the object? It's fairly easy to marshall/unmarshall into a binary format which acts as a deep copy, but it would be nice to only need to do that when the object is retrieved from the cache

Reply:
I don't know of a way to do this with the annotations, it might be easier in this case to use the raw cache API:

```java
@Autowired
private CacheManager cacheManager;

...

cache = cacheManager.getCache("case_search")
value = cache.get(key, valueLoader)
...
cache.put(key, serializedValue)
```

Reply:
@ctsims I am having trouble serializing `ExternalDataInstance` correctly, It seems to skip the [root node](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/javarosa/core/model/instance/ExternalDataInstance.java#L19) during [serialization](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/javarosa/core/model/instance/ExternalDataInstance.java#L99) today as root type `AbstractTreeElement` is an interface here. I am not sure what's the best way forward to serialize this. 

Also I did survey ExternalDataInstance and it's containing objects and they all are completely immutable.

Reply:
Ah, I see what's going on. I mistook what the serialization of `ExternalDataInstance` actually accomplished.

It looks like `ExternalDataInstance`'s serialization is used to store **a placeholder** for a data instance that is connected up at runtime and that when we serialize concrete DOM's (like an inline instance in a form) we end up parsing them as [FormInstance](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/javarosa/xform/parse/XFormParser.java#L541) objects.

Options moving forward:

*One:* You could instead switch to serializing the [TreeEleement Root](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/commcare/session/RemoteQuerySessionManager.java#L131) itself, since that's how External Data Instance is actually created. `TreeElement`s should be reliably portable. 

*Two:* Given the added complexity of the serialized deep copy, I think it sounds fine to start with the assumption that we'll cache the in-memory version. It does look like the underlying structure is coded with a `TreeElement` DOM which is mutable and has [at least one](https://github.com/dimagi/commcare-core/blob/b47f46d332abd402ba7833dd6617f03306272bf9/src/main/java/org/javarosa/core/model/instance/TreeElement.java#L81) synchronicity concern, but since the DOM is used immutably and doesn't change after init I think it won't be an issue.